README.md
# SwipeUpView

[![CI Status](https://img.shields.io/travis/kadirkemal/SwipeUpView.svg?style=flat)](https://travis-ci.org/kadirkemal/SwipeUpView)
[![Version](https://img.shields.io/cocoapods/v/SwipeUpView.svg?style=flat)](https://cocoapods.org/pods/SwipeUpView)
[![License](https://img.shields.io/cocoapods/l/SwipeUpView.svg?style=flat)](https://cocoapods.org/pods/SwipeUpView)
[![Platform](https://img.shields.io/cocoapods/p/SwipeUpView.svg?style=flat)](https://cocoapods.org/pods/SwipeUpView)

## SCREENSHOT

<img src="https://raw.githubusercontent.com/zingat/SwipeUpView/master/SwipeUpView.gif" width="200px">

## Installation

SwipeUpView is available through [CocoaPods](http://cocoapods.org). To install it, simply add the following line to your Podfile:

```
pod 'SwipeUpView'
```

## SwipeUpViewDatasource
```objectivec
extension ViewController : SwipeUpViewDatasource {

    func hideHeaderButton(_ swipeUpView: SwipeUpView) -> Bool {
        return false
    }

    func heightOfHeaderButton (_ swipeUpView : SwipeUpView) -> CGFloat {
        return 8.0
    }

    func widthOfHeaderButton (_ swipeUpView : SwipeUpView) -> CGFloat {
        return 50.0
    }

    func marginOfHeaderButton (_ swipeUpView : SwipeUpView) -> CGFloat {
        return 4.0
    }

    func colorOfHeaderButton (_ swipeUpView : SwipeUpView) -> UIColor {
        return .white
    }

    func firstOpenHeightIndex(_ swipeUpView: SwipeUpView) -> Int {
        return 1
    }

    ////SETTING HEIGHTS
    //if heights returns a non-empty CGFloat array, heightPercentages return value is not important
    func heights(_ swipeUpView: SwipeUpView) -> [CGFloat] {
        return [100, 300, 600]
    }
    
    func heightPercentages(_ swipeUpView: SwipeUpView) -> [CGFloat] {
        return [0.1, 0.5, 0.95]
    }
    ////

    ////SETTING HEIGHTS - 2
    //if heights returns an empty CGFloat array, heightPercentages return value will be used
    func heights(_ swipeUpView: SwipeUpView) -> [CGFloat] {
        return []
    }

    func heightPercentages(_ swipeUpView: SwipeUpView) -> [CGFloat] {
        return [0.1, 0.5, 0.95]
    }
    ////

}
```

## SwipeUpViewDelegate
```objectivec
extension ViewController : SwipeUpViewDelegate {

    func swipeUpViewStateWillChange (_ swipeUpView : SwipeUpView, stateIndex : Int){
        NSLog("SwipeUpView state will change to %i", stateIndex)
    }

    func swipeUpViewStateDidChange (_ swipeUpView : SwipeUpView, stateIndex : Int){
        NSLog("SwipeUpView state did change to %i", stateIndex)
    }

    func swipeUpViewWillOpen (_ swipeUpView : SwipeUpView){
        NSLog("SwipeUpView state will open")
    }

    func swipeUpViewDidOpen (_ swipeUpView : SwipeUpView){
        NSLog("SwipeUpView state did open")
    }

    func swipeUpViewWillClose (_ swipeUpView : SwipeUpView){
        NSLog("SwipeUpView state will close")
    }

    func swipeUpViewDidClose (_ swipeUpView : SwipeUpView){
        NSLog("SwipeUpView state did close")
    }

}
```

## Authors

Zingat Mobile Team
+ Yusuf Çınar, https://github.com/cinaryusufiu
+ Kadir Kemal Dursun, https://github.com/KadirKemal

## License

SwipUpView is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log

## [v1.2.0] (https://github.com/zingat/SwipeUpView/releases/tag/1.2.0)
- These methods were added. 
  - goToUp(stepCount : Int)
  - goToTop()
  - goToDown(stepCount : Int)
  - goToBottom()

## [v1.1.0] (https://github.com/zingat/SwipeUpView/releases/tag/1.1.0)
- Adding static heights 

## [v1.0.1] (https://github.com/zingat/SwipeUpView/releases/tag/1.0.1)
- Adding @objc in protocols definition 

## [v1.0.0] (https://github.com/zingat/SwipeUpView/releases/tag/1.0.0)
- Hello World



Example/Tests/Tests.swift
import XCTest
import SwipeUpView

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Example/SwipeUpView/BottomView.swift
//
//  BottomView.swift
//  SwipeUpView_Example
//
//  Created by Kadir Kemal Dursun on 29.06.2018.
//  Copyright © 2018 CocoaPods. All rights reserved.
//

import UIKit

public class BottomView: UIView  {
    
    @IBOutlet var contentView: UIView!
    @IBOutlet weak var imageView: UIImageView!
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInit()
    }
    
    private func commonInit(){
        Bundle.main.loadNibNamed("BottomView", owner: self, options: nil)
        addSubview(contentView)
        contentView.frame = self.bounds
        contentView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    }
    
    public func useVerticalImage() {
        self.imageView.image = UIImage(named: "Zingat_dikey_logo")
    }
    
    public func useHorizontalImage() {
        self.imageView.image = UIImage(named: "zingat_yatay_logo")
    }
    
    
    
    
}


Example/SwipeUpView/AppDelegate.swift
//
//  AppDelegate.swift
//  SwipeUpView
//
//  Created by kadirkemal on 06/29/2018.
//  Copyright (c) 2018 kadirkemal. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



SwipeUpView/Classes/SwipeUpView.swift
//
//  SwipeUpView.swift
//  Pods-SwipeUpView_Example
//
//  Created by cinaryusuf on 28.06.2018.
//

import UIKit

public class SwipeUpView: UIView  {
  
    private var isHeaderButtonDirectionToTop = true
    private var activeIndex = 0
    private var isOpen = false
    
    private weak var mainView : UIView?
    
    public weak var delegate: SwipeUpViewDelegate?
    public weak var datasource : SwipeUpViewDatasource?
    
    
    public init(frame: CGRect, mainView: UIView){
        super.init(frame: frame)
    
        self.isUserInteractionEnabled = true
        self.backgroundColor = .clear
        self.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(onClickedHeaderButton)))
        
        self.mainView = mainView
        self.frame = CGRect(x: 0, y: mainView.frame.height , width: mainView.frame.width, height: 0);
    }
    
    
    public var swipeContentView : UIView? {
        didSet {
            addItemContainerView()
        }
    }
    
    
    public lazy var headerContainerButton : UIButton = {
        
        let button = UIButton(type: UIButtonType.system)
        button.setTitle("", for: UIControlState.normal)
        button.clipsToBounds = true
        
        if let datasource = self.datasource {
            button.backgroundColor = datasource.colorOfHeaderButton(self)
        }else{
            button.backgroundColor = UIColor.darkGray
        }
        
        button.layer.cornerRadius = 4
        button.addTarget(self, action: #selector(onClickedHeaderButton), for: .touchUpInside)
        return button
        
    }()
    
    lazy var headerButtonPanGesture : UIPanGestureRecognizer = {
        return UIPanGestureRecognizer(target: self, action: #selector(handleButtonPanGesture(recognizer:)))
    }()
    
    @objc func handleButtonPanGesture(recognizer : UIPanGestureRecognizer){
        
        guard let mainView = self.mainView else {  return }
        
        if recognizer.state == .ended {
            let index = findNewHeightPercentageIndex()
            adjustHeaderButtonDirection(index: index)
            self.adjustMyHeight(heigthIndex: index);
        }
        
        //mainView.bringSubview(toFront: self)
        
        let translationRec = recognizer.translation(in: mainView)
        self.center = CGPoint(x: self.center.x , y: self.center.y + translationRec.y)
        
        var f = self.frame;
        f.size.height = mainView.frame.height - f.origin.y;

        self.frame = f;
        
        recognizer.setTranslation(.zero, in: mainView)
    }
    
    lazy var swipeGestureRecognizerDown : UISwipeGestureRecognizer = {
        let  sg = UISwipeGestureRecognizer(target: self, action: #selector(handleSwipeGestureDown(sender :)))
        sg.direction = .down
        
        return sg
    }()
    lazy var swipeGestureRecognizerUp : UISwipeGestureRecognizer = {
        let  sg = UISwipeGestureRecognizer(target: self, action: #selector(handleSwipeGestureDown(sender :)))
        sg.direction = .down
        sg.numberOfTouchesRequired = 1
        return sg
    }()
    
    
    
    @objc func handleSwipeGestureDown(sender : UISwipeGestureRecognizer)  {
        adjustMyHeight(heigthIndex: setupNewIndexForIndex(index: 0))
        isHeaderButtonDirectionToTop = true
    }
    @objc func handleSwipeGestureUp(sender : UISwipeGestureRecognizer)  {
        guard  let datasource = self.datasource else { return }
        
        let stateHeightPercentages = datasource.heightPercentages(self)
        
        adjustMyHeight(heigthIndex: setupNewIndexForIndex(index: stateHeightPercentages.count - 1))
        isHeaderButtonDirectionToTop = false
    }
    
    
    private func findNewHeightPercentageIndex() -> Int{
        guard let mainView = self.mainView, let datasource = self.datasource else {
            return 0
        }
        
        if(datasource.heights(self).count > 0){
            return findNewHeightIndex();
        }
        
        let ratio = self.frame.height / mainView.frame.height;
        
        let stateHeightPercentages = datasource.heightPercentages(self)
        let activeRatio = stateHeightPercentages[activeIndex]
        
        if(ratio < activeRatio && activeIndex > 0){
            for index in (0 ... activeIndex-1).reversed() {
                if stateHeightPercentages[index] < ratio{
                    return index
                }
            }
            
            return 0
        }
        
        if(ratio > activeRatio && activeIndex < stateHeightPercentages.count - 1){
            for index in activeIndex+1 ..< stateHeightPercentages.count{
                if stateHeightPercentages[index] > ratio{
                    return index
                }
            }
            
            return stateHeightPercentages.count - 1
        }

        return activeIndex
    }
    
    private func findNewHeightIndex() -> Int{
        guard let datasource = self.datasource else {
            return 0
        }
        
        let currentHight = self.frame.height;
        
        let stateHeights = datasource.heights(self)
        let activeHeight = stateHeights[activeIndex]
        
        if(currentHight < activeHeight && activeIndex > 0){
            for index in (0 ... activeIndex-1).reversed() {
                if stateHeights[index] < currentHight{
                    return index
                }
            }
            
            return 0
        }
        
        if(currentHight > activeHeight && activeIndex < stateHeights.count - 1){
            for index in activeIndex+1 ..< stateHeights.count{
                if stateHeights[index] > currentHight{
                    return index
                }
            }
            
            return stateHeights.count - 1
        }
        
        return activeIndex
    }
    
    @objc func onClickedHeaderButton(){
        let index = findNewHeightPercentageIndex()
        adjustHeaderButtonDirection(index: index)
        adjustMyHeight(heigthIndex: setupNewIndexForIndex(index: index))
    }
    
    private func adjustMyHeight(heigthIndex : Int){
        guard let mainView = self.mainView, let datasource = self.datasource  else { return  }
        
        self.delegate?.swipeUpViewStateWillChange(self, stateIndex: heigthIndex)
        
        activeIndex = heigthIndex
        
        let heightPercentages = datasource.heightPercentages(self)
        let heights = datasource.heights(self)
        
        var newHeight: CGFloat = 0.0;
        if(heights.count > 0){
            newHeight = heights[heigthIndex];
        }else{
            newHeight = mainView.frame.size.height * heightPercentages[heigthIndex];
        }
        
        
        UIView.animate(withDuration: 0.7, delay: 0, usingSpringWithDamping: 0.6, initialSpringVelocity: 0.8, options: .curveEaseInOut, animations: {
            var frame = self.frame
            frame.origin.y = mainView.frame.size.height - newHeight
            frame.size.height = newHeight
            self.frame = frame
            
            self.layoutIfNeeded()
        }, completion: { (_) in
            self.delegate?.swipeUpViewStateDidChange(self, stateIndex: heigthIndex)
            if(self.isOpen == false){
                self.isOpen = true
                self.delegate?.swipeUpViewDidOpen(self)
            }
        });
        
    }
    private func setupNewIndexForIndex(index : Int) -> Int {
        return isHeaderButtonDirectionToTop ? index + 1 : index - 1
    }
    
    private func addItemContainerView() {
        guard let swipeContentView = self.swipeContentView , let mainView = self.mainView else { return  }
        
        let width = mainView.frame.size.width
        swipeContentView.layer.cornerRadius = 6
        
        self.addGestureRecognizer(swipeGestureRecognizerUp)
        self.addGestureRecognizer(swipeGestureRecognizerDown)
        
        self.addSubview(headerContainerButton)
        self.addSubview(swipeContentView)
        
        self.addGestureRecognizer(self.headerButtonPanGesture)
        
        frameItemContainerView(width: width)
    }
    
    
    private  func frameItemContainerView(width : CGFloat) {
        guard let datasource = self.datasource , let swipeContentView = self.swipeContentView  else { return  }
        
        var contentX : CGFloat = 0.0
        if datasource.hideHeaderButton(self) == false {
            contentX = datasource.heightOfHeaderButton(self) + datasource.marginOfHeaderButton(self)
            
            self.headerContainerButton.translatesAutoresizingMaskIntoConstraints = false
            self.addConstraints([
                NSLayoutConstraint(item: self.headerContainerButton, attribute: .centerX, relatedBy: .equal, toItem: self, attribute: .centerX, multiplier: 1.0, constant: 0),
                NSLayoutConstraint(item: self.headerContainerButton, attribute: .height, relatedBy: .equal, toItem: nil, attribute: .height, multiplier: 1.0, constant: datasource.heightOfHeaderButton(self)),
                NSLayoutConstraint(item: self.headerContainerButton, attribute: .width, relatedBy: .equal, toItem: nil, attribute: .width, multiplier: 1.0, constant: datasource.widthOfHeaderButton(self)),
                NSLayoutConstraint(item: self.headerContainerButton, attribute: .top, relatedBy: .equal, toItem: self, attribute: .top, multiplier: 1.0, constant: 0)
                ])
        }
        
        swipeContentView.frame.size.width = self.frame.size.width
        swipeContentView.translatesAutoresizingMaskIntoConstraints = false
        self.addConstraints([
            NSLayoutConstraint(item: swipeContentView, attribute: .top, relatedBy: .equal, toItem: self, attribute: .top, multiplier: 1.0, constant: contentX),
            NSLayoutConstraint(item: swipeContentView, attribute: .bottom, relatedBy: .equal, toItem: self, attribute: .bottom, multiplier: 1.0, constant: 0),
            NSLayoutConstraint(item: swipeContentView, attribute: .leading, relatedBy: .equal, toItem: self, attribute: .leading, multiplier: 1.0, constant: 0),
            NSLayoutConstraint(item: swipeContentView, attribute: .trailing, relatedBy: .equal, toItem: self, attribute: .trailing, multiplier: 1.0, constant: 0),
            ])
        
    }
    
    public func openViewPage()  {
        
        guard let mainView = self.mainView, let datasource = self.datasource else {
            return
        }
        self.delegate?.swipeUpViewWillOpen(self)
        
        mainView.addSubview(self)
        self.frame = CGRect(x: 0, y: mainView.frame.height , width: mainView.frame.width, height: 0);
        
        adjustMyHeight(heigthIndex: datasource.firstOpenHeightIndex(self))
    }
    
    public  func closeViewPage()  {
        
        guard let mainView = self.mainView else { return  }
        
        self.delegate?.swipeUpViewWillClose(self)
        UIView.animate(withDuration: 0.3, delay: 0, usingSpringWithDamping: 0.6, initialSpringVelocity: 0.8, options: .curveEaseInOut, animations: {
            
            self.frame = CGRect(x: 0, y: mainView.frame.height , width: mainView.frame.width, height: mainView.frame.height);
            
        }) { (_) in
            self.delegate?.swipeUpViewDidClose(self)
            self.isOpen = false
            self.removeFromSuperview()
        }
    }
    
    func heightCount() -> Int {
        guard let datasource = self.datasource else {
            return 0
        }
        
        let stateHeightPercentages = datasource.heightPercentages(self)
        let stateHeights = datasource.heights(self)
        
        let hCount = stateHeights.count
        if(hCount > 0) {
            return hCount
        }
        
        return stateHeightPercentages.count
    }
    
    func adjustHeaderButtonDirection(index : Int){
        let hCount = heightCount();
        
        if index == hCount - 1 {
            self.isHeaderButtonDirectionToTop = false
        }
        if  index == 0 {
            self.isHeaderButtonDirectionToTop = true
        }
    }
    
    public func goToUp(stepCount : Int){
        let hCount = heightCount();
        var sCount = stepCount
        if(activeIndex + sCount >= hCount){
            sCount = hCount - activeIndex - 1
        }
        if(sCount < 0){
            return
        }
        
        self.adjustMyHeight(heigthIndex: activeIndex + sCount);
    }
    
    public func goToTop(){
        let hCount = heightCount();
        self.adjustMyHeight(heigthIndex: hCount - 1);
    }
    
    public func goToDown(stepCount : Int){
        var sCount = stepCount
        
        if(activeIndex - sCount < 0){
            sCount = activeIndex
        }
        
        self.adjustMyHeight(heigthIndex: activeIndex - sCount);
    }
    
    public func goToBottom(){
        self.adjustMyHeight(heigthIndex: 0);
    }
    
    
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}



Example/SwipeUpView/ViewController.swift
//
//  ViewController.swift
//  SwipeUpView
//
//  Created by kadirkemal on 06/29/2018.
//  Copyright (c) 2018 kadirkemal. All rights reserved.
//

import UIKit
import SwipeUpView

class ViewController: UIViewController {
    var isOpenCloseControl = true
    
    @IBOutlet weak var buttonSet: UIStackView!
    
    lazy var bottomView : BottomView = {
        return BottomView(frame:.zero)
    }()
    
    lazy var swipeUpView : SwipeUpView = {
        let mg = SwipeUpView(frame: .zero, mainView: self.view)
        mg.delegate = self
        mg.datasource = self
        
        mg.swipeContentView = bottomView
        
        return mg
    }()
    
    @IBAction func onClickedButton(_ sender: Any) {
        if isOpenCloseControl  {
            isOpenCloseControl = false
            swipeUpView.openViewPage()
            
        }
        else {
            isOpenCloseControl = true
            swipeUpView.closeViewPage()
            
        }
    }
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        buttonSet.isHidden = true
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func onClickedUpButton(_ sender: Any) {
        swipeUpView.goToUp(stepCount: 1)
    }
    
    @IBAction func onClickedDownButton(_ sender: Any) {
        swipeUpView.goToDown(stepCount: 1)
    }
    
    @IBAction func onClickedTopButton(_ sender: Any) {
        swipeUpView.goToTop()
    }
    
    @IBAction func onClickedBottomButton(_ sender: Any) {
        swipeUpView.goToBottom()
    }
    
    
}

extension ViewController : SwipeUpViewDatasource ,SwipeUpViewDelegate {
    
    func hideHeaderButton(_ swipeUpView: SwipeUpView) -> Bool {
        return false
    }
    
    func heightOfHeaderButton (_ swipeUpView : SwipeUpView) -> CGFloat {
        return 8.0
    }
    
    func widthOfHeaderButton (_ swipeUpView : SwipeUpView) -> CGFloat {
        return 50.0
    }
    
    func marginOfHeaderButton (_ swipeUpView : SwipeUpView) -> CGFloat {
        return 4.0
    }
    
    func colorOfHeaderButton (_ swipeUpView : SwipeUpView) -> UIColor {
        return .white
    }
    
    func firstOpenHeightIndex(_ swipeUpView: SwipeUpView) -> Int {
        return 1
    }
    
    func heights(_ swipeUpView: SwipeUpView) -> [CGFloat] {
        return [100, 300, 600]
    }
    
    func heightPercentages(_ swipeUpView: SwipeUpView) -> [CGFloat] {
        return [0.1, 0.5, 0.95]
    }
    
    func swipeUpViewStateWillChange (_ swipeUpView : SwipeUpView, stateIndex : Int){
        NSLog("SwipeUpView state will change to %i", stateIndex)
        
        if stateIndex == 2 {
            bottomView.useVerticalImage()
        }else{
            bottomView.useHorizontalImage()
        }
    }
    
    func swipeUpViewStateDidChange (_ swipeUpView : SwipeUpView, stateIndex : Int){
        NSLog("SwipeUpView state did change to %i", stateIndex)
    }
    
    func swipeUpViewWillOpen (_ swipeUpView : SwipeUpView){
        NSLog("SwipeUpView state will open")
        buttonSet.isHidden = false
    }
    
    func swipeUpViewDidOpen (_ swipeUpView : SwipeUpView){
        NSLog("SwipeUpView state did open")
    }
    
    func swipeUpViewWillClose (_ swipeUpView : SwipeUpView){
        NSLog("SwipeUpView state will close")
        buttonSet.isHidden = true
    }
    
    func swipeUpViewDidClose (_ swipeUpView : SwipeUpView){
        NSLog("SwipeUpView state did close")
    }
    
    
    
}
